### PR TITLE
Grouped format implementation

### LINKED ISSUE #948: feat: Alternative Text Output Format(s)
I would like to group the errors in ruff by file and output them in a more structured format Similar to the image below where issues are grouped by file and then displayed in a table like view below the file name. 

**Example**

CleanShot 2022-11-28 at 13 15 02@2x

Obviously this is mostly a personal preference, but I find that much easier to read than the current output. 

I could see this being implement in a two different ways.

1. Add an additional option to `--format` flag that displays an output like above. This is likely the easiest option
2. Accept a template of some kind in either the config file or as a parameter to --format.  Perhaps with a template engine like [tera](https://github.com/Keats/tera). This is the most flexible, but difficult and would add additional dependencies

I'm happy to work on a PR for this if it's a feature you're interest in including in ruff. 

Thanks for all your work on this!

### PR DESCRIPTION
**Pull request by hay-kot**
This PR adds an additional option to the `--format` flag "grouped" which changes the output. Things to note

1. I extracted the pre and post printing parts into their own functions
2. This implementation relies on the diagnostic messages to be sorted by file
3. I don't  know Rust that well so please let me know if anything needs fixing! 

Closes #948 

**Output Example**

CleanShot 2022-11-28 at 21 33 41@2x

### CODE REVIEW

**Comment by charliermarsh** [comment]
Awesome, thank you! Will review today.

File: src/printer.rs [hunk 1]
```diff
@@ -41,6 +45,28 @@ impl<'a> Printer<'a> {
...
+                    diagnostics.messages.len(),
+                    diagnostics.fixed,
+                );
+            } else if !diagnostics.messages.is_empty() {
+                println!("Found {} error(s).", diagnostics.messages.len());
+            }
+        }
+    }
+
+    fn post_test(&self, num_fixable: usize) {
```
Comment (charliermarsh) [inline src/printer.rs:62]: Nit: small typo (`post_text`)

File: src/printer.rs [hunk 2]
```diff
@@ -73,27 +99,41 @@ impl<'a> Printer<'a> {
...
-                    if num_fixable > 0 {
-                        println!("{num_fixable} potentially fixable with the --fix option.");
+                self.post_test(num_fixable);
+            }
+            SerializationFormat::Grouped => {
+                self.pre_text(diagnostics);
+
+                let mut filename = String::new();
+
+                for message in &diagnostics.messages {
```
Comment (charliermarsh) [inline src/printer.rs:115]: I think this assumes that the messages are sorted by filename. Maybe slightly preferable to create a HashMap from filename to messages? That can also help us avoid the clones.

Comment (charliermarsh) [reply src/printer.rs:115]: Ehh... maybe this is simpler, actually.

**Comment by charliermarsh** [comment]
Looks good to me! I'm going to make some minor tweaks to the output formatting, then merge.

File: src/printer.rs (hunk 2 above)
Comment (hay-kot) [reply src/printer.rs:115]: I took a stab at a hashmap for the sake of it. I can't really make an educated guess on perf, but I think either is simple enough to understand. 🤷 

```rust
let mut map = std::collections::HashMap::new();

for message in &diagnostics.messages {
    map.entry(&message.filename)
        .or_insert_with(Vec::new)
        .push(message);
}

for (filename, messages) in map {
    println!(
        "\n{}:",
        relativize_path(Path::new(&filename)).bold().underline()
    );
    for message in messages {
        println!(
            "    {}{}{} {}  {}",
            message.location.row(),
            ":".cyan(),
            message.location.column(),
            message.kind.code().as_ref().red().bold(),
            message.kind.body(),
        );
    }
}
```

**Comment by charliermarsh** [comment]
(Namely, I'm going to change it to mirror ESLint's behavior, such that the `row:column` column is always aligned, and there are always two spaces between each section.)

**Comment by charliermarsh** [comment]
Thank you!

**Comment by charliermarsh** [comment]
We should make `format` a supported option in `pyproject.toml` too.

**Pull request merged**